### PR TITLE
github: Add workflow for deploying a CCv0 demo

### PR DESCRIPTION
**Pull request by Jakob-Naucke**
docs: Create sample config for confidential agent

Basic config, no debug endpoints, no exec/reseed. Uses the
`$AA_KBC_PARAMS` variable to be used with `envsubst`.

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

---

github: Add workflow for deploying a CCv0 demo

using the offline FS KBC [1] and keys from the SSH demo [2]. The
workflow is adapted from `main:kata-deploy-test.yaml`. The image
deployed here is _not_ for a trusted execution environment.

[1] - https://github.com/confidential-containers/attestation-agent/tree/main/src/kbc_modules/offline_fs_kbc
[2] - https://github.com/confidential-containers/documentation/tree/main/demos/ssh-demo

Fixes: #3198
Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

---

Here is a diff to `kata-deploy-test.yaml` to make this PR a little more comprehensible:

```diff
@@ -2,7 +2,7 @@ on:
   issue_comment:
     types: [created, edited]
 
-name: test-kata-deploy
+name: deploy-ccv0-demo
 
 jobs:
   check-comment-and-membership:
@@ -11,7 +11,7 @@ jobs:
       github.event.issue.pull_request
       && github.event_name == 'issue_comment'
       && github.event.action == 'created'
-      && startsWith(github.event.comment.body, '/test_kata_deploy')
+      && startsWith(github.event.comment.body, '/deploy-ccv0-demo')
     steps:
       - name: Check membership
         uses: kata-containers/is-organization-member@1.0.1
@@ -54,6 +54,14 @@ jobs:
           curl -fsSL https://test.docker.com -o test-docker.sh
           sh test-docker.sh
 
+      - name: Prepare confidential container rootfs
+        if: ${{ matrix.asset == 'rootfs-initrd' }}
+        run: |
+          wget -P include_rootfs/etc/ https://raw.githubusercontent.com/confidential-containers/documentation/main/demos/ssh-demo/aa-offline_fs_kbc-keys.json
+          envsubst < docs/how-to/data/confidential-agent-config.toml.in > include_rootfs/etc/confidential-agent-config.toml
+        env:
+          AA_KBC_PARAMS: offline_fs_kbc::null
+
       - name: Build ${{ matrix.asset }}
         run: |
           make "${KATA_ASSET}-tarball"
@@ -61,6 +69,8 @@ jobs:
           # store-artifact does not work with symlink
           sudo cp -r "${build_dir}" "kata-build"
         env:
+          AA_KBC: offline_fs_kbc
+          INCLUDE_ROOTFS: include_rootfs
           KATA_ASSET: ${{ matrix.asset }}
           TAR_OUTPUT: ${{ matrix.asset }}.tar.gz
 
@@ -108,19 +118,9 @@ jobs:
           pkg_sha=$(git rev-parse HEAD)
           popd
           mv kata-static.tar.xz $GITHUB_WORKSPACE/tools/packaging/kata-deploy/kata-static.tar.xz
-          docker build --build-arg KATA_ARTIFACTS=kata-static.tar.xz -t quay.io/kata-containers/kata-deploy-ci:$pkg_sha $GITHUB_WORKSPACE/tools/packaging/kata-deploy
+          docker build --build-arg KATA_ARTIFACTS=kata-static.tar.xz -t quay.io/confidential-containers/kata-demo:$pkg_sha $GITHUB_WORKSPACE/tools/packaging/kata-deploy
           docker login -u ${{ secrets.QUAY_DEPLOYER_USERNAME }} -p ${{ secrets.QUAY_DEPLOYER_PASSWORD }} quay.io
-          docker push quay.io/kata-containers/kata-deploy-ci:$pkg_sha
+          docker push quay.io/confidential-containers/kata-demo:$pkg_sha
           mkdir -p packaging/kata-deploy
           ln -s $GITHUB_WORKSPACE/tools/packaging/kata-deploy/action packaging/kata-deploy/action
           echo "::set-output name=PKG_SHA::${pkg_sha}"
-      - name: test-kata-deploy-ci-in-aks
-        uses: ./packaging/kata-deploy/action
-        with:
-          packaging-sha: ${{steps.build-and-push-kata-deploy-ci.outputs.PKG_SHA}}
-        env:
-          PKG_SHA: ${{steps.build-and-push-kata-deploy-ci.outputs.PKG_SHA}}
-          AZ_APPID: ${{ secrets.AZ_APPID }}
-          AZ_PASSWORD: ${{ secrets.AZ_PASSWORD }}
-          AZ_SUBSCRIPTION_ID: ${{ secrets.AZ_SUBSCRIPTION_ID }}
-          AZ_TENANT_ID: ${{ secrets.AZ_TENANT_ID }}
```

/cc @fidencio @sameo -- I don't think I got everything about the workflow, but I'd rather start with _something_. Can I just `/deploy-ccv0-demo` from here?
/cc @bpradipt -- I think it would be good to integrate this with https://github.com/confidential-containers/operator/tree/ccv0-demo/demo/payload-image

### CODE REVIEW

File: .github/workflows/deploy-ccv0-demo.yaml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+name: deploy-ccv0-demo
+
+jobs:
+  check-comment-and-membership:
+    runs-on: ubuntu-latest
+    if: |
+      github.event.issue.pull_request
+      && github.event_name == 'issue_comment'
+      && github.event.action == 'created'
+      && startsWith(github.event.comment.body, '/test_kata_deploy')
```
Comment (fidencio) [inline .github/workflows/deploy-ccv0-demo.yaml:14]: @Jakob-Naucke, this is the comment that will trigger the action.  You'll need to change that to `/deploy-ccv0-demo` (or whatever you want to call).

File: .github/workflows/deploy-ccv0-demo.yaml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      - uses: actions/checkout@v2
+      - name: Install docker
+        run: |
+          curl -fsSL https://test.docker.com -o test-docker.sh
+          sh test-docker.sh
+
+      - name: Build ${{ matrix.asset }}
+        run: |
+          wget -P include_rootfs/etc/ https://raw.githubusercontent.com/confidential-containers/documentation/main/demos/ssh-demo/aa-offline_fs_kbc-keys.json
+          envsubst docs/how-to/data/confidential-agent-config.toml.in > include_rootfs/etc/confidential-agent-config.toml
```
Comment (fidencio) [inline .github/workflows/deploy-ccv0-demo.yaml:60]: I'm not sure if this is the right placeto add those, sincerely.
Have you tested it?  Does it produce the results you want?

File: .github/workflows/deploy-ccv0-demo.yaml [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      - name: test-kata-deploy-ci-in-aks
+        uses: ./packaging/kata-deploy/action
+        with:
+          packaging-sha: ${{steps.build-and-push-kata-deploy-ci.outputs.PKG_SHA}}
+        env:
+          PKG_SHA: ${{steps.build-and-push-kata-deploy-ci.outputs.PKG_SHA}}
+          AZ_APPID: ${{ secrets.AZ_APPID }}
+          AZ_PASSWORD: ${{ secrets.AZ_PASSWORD }}
+          AZ_SUBSCRIPTION_ID: ${{ secrets.AZ_SUBSCRIPTION_ID }}
+          AZ_TENANT_ID: ${{ secrets.AZ_TENANT_ID }}
```
Comment (fidencio) [inline .github/workflows/deploy-ccv0-demo.yaml:131]: Do you want this to happen?  Do you want to test this on AKS or do you just want to upload the image to quay.io?

I don't think testing on AKS is worth it, as that won't have the containerd version required by the CCv0 testing, but you let me know about your plans.

**Review by fidencio — COMMENTED** [review]
Apart from a comment made in the code about changing the "action name",  I'd like to let you know that in order to actually call and test the action, you need to have it merged first.  So, I'd recommend you merge it on your repo, and test it from there (even if you don't have the AKS permissions, you'd be able to catch errors / inconsistencies by testing on your own repo).

There are more comments in the PR, inline.

File: .github/workflows/deploy-ccv0-demo.yaml (hunk 2 above)
Comment (Jakob-Naucke) [reply .github/workflows/deploy-ccv0-demo.yaml:60]: > Have you tested it? Does it produce the results you want?

No, I have not tested this yet; I wouldn't have known how.

> So, I'd recommend you merge it on your repo, and test it from there

Thanks! I will.

**Comment by Jakob-Naucke** [comment]
There will be some updates required (still coming) but I have a question already. I have found (as you have already alluded @fidencio) that actions must be in the default branch to work, which won't work with kata-containers/kata-containers/CCv0. I see pretty much two paths going forward:
1. Merge this into kata-containers/kata-containers/main.
2. Run these actions on confidential-containers/kata-containers/CCv0 (which is the default there), which was supposed to be a read-only mirror. (We wouldn't have code that's any different there, but we would run the actions.)

I prefer (2). @sameo WDYT?

**Comment by Jakob-Naucke** [comment]
not tested e2e yet but I looked at an initrd and it had the respective things, moving out of draft

**Comment by fidencio** [comment]
> There will be some updates required (still coming) but I have a question already. I have found (as you have already alluded @fidencio) that actions must be in the default branch to work, which won't work with kata-containers/kata-containers/CCv0. I see pretty much two paths going forward:
> 
>     1. Merge this into kata-containers/kata-containers/main, even though this depends on code in the CCv0 branch.
> 
>     2. Run these actions on confidential-containers/kata-containers/CCv0 (which is the default there), which was supposed to be a read-only mirror. (We wouldn't have code that's any different there, but we would run the actions.)
> 
> 
> I prefer (2). @sameo WDYT?

Not @sameo here, but let me nosy and give my opinion. :-)
I don't mind going for (2) as long as we ensure we'll have that merged back to the main kata-containers branch at some point in the future (as in, we need to merge that back when we start merging the CCv0 content into main).  With this in mind, maybe (1) would be less work in the future?

**Comment by Jakob-Naucke** [comment]
GitHub's "quote reply" is broken and I have to copy and paste _manually_...

> I don't mind going for (2) as long as we ensure we'll have that merged back to the main kata-containers branch at some point in the future (as in, we need to merge that back when we start merging the CCv0 content into main). With this in mind, maybe (1) would be less work in the future?

I don't really think it's extra work, it's a new file, I don't see any potential for merge conflicts.

**Comment by Jakob-Naucke** [comment]
okay yeah this works e2e /test

**Comment by Jakob-Naucke** [comment]
/retest-fedora
for reference:
```
kube-system   kube-apiserver-fedoradocker32-923340            1/1     Running   0          7s
kube-system   kube-controller-manager-fedoradocker32-923340   1/1     Running   0          7s
kube-system   etcd-fedoradocker32-923340                      1/1     Running             0          17s
kube-system   kube-scheduler-fedoradocker32-923340            1/1     Running             0          16s
[run_kubernetes_tests.sh:68] ERROR: ./init.sh
```

which is usually disk pressure

**Comment by Jakob-Naucke** [comment]
Naming config file `kata-config.toml` instead of `confidential-agent-config.toml` because I've called it that elsewhere, let's keep it consistent

**Comment by Jakob-Naucke** [comment]
@bpradipt this would be my logic for the config. I have tested the changes, but not end-to-end. We should decide on where we should set it.
```patch
From 974d6b29af3dd5d6860282cb09c4998032ab4b42 Mon Sep 17 00:00:00 2001
From: Jakob Naucke <jakob.naucke@ibm.com>
Date: Fri, 10 Dec 2021 16:44:57 +0100
Subject: [PATCH] kata-deploy: Configure Kata & containerd for CCv0

- Specify cri_handler in containerd config
- Specify to use initrd
- Specify kernel_params according to guest config
- Specify service_offload

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>
---
 .../kata-deploy/kata-deploy/base/kata-deploy.yaml  |  2 ++
 tools/packaging/kata-deploy/scripts/kata-deploy.sh | 14 +++++++++++++-
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/tools/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml b/tools/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml
index d500b1f1f..3e2aba340 100644
--- a/tools/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml
+++ b/tools/packaging/kata-deploy/kata-deploy/base/kata-deploy.yaml
@@ -28,6 +28,8 @@ spec:
           valueFrom:
             fieldRef:
               fieldPath: spec.nodeName
+        - name: CONFIGURE_CC
+          value: "yes"
         securityContext:
           privileged: false
         volumeMounts:
diff --git a/tools/packaging/kata-deploy/scripts/kata-deploy.sh b/tools/packaging/kata-deploy/scripts/kata-deploy.sh
index abdfc6777..11bba0ce6 100755
--- a/tools/packaging/kata-deploy/scripts/kata-deploy.sh
+++ b/tools/packaging/kata-deploy/scripts/kata-deploy.sh
@@ -171,7 +171,8 @@ function configure_containerd_runtime() {
 	else
 		cat <<EOT | tee -a "$containerd_conf_file"
 [$runtime_table]
-  runtime_type = "${runtime_type}"
+  runtime_type = "${runtime_type}" \
+$([ "$runtime" == "kata" ] || [ "$runtime" == "kata-qemu" ] && [ "${CONFIGURE_CC:-}" == "yes" ] && printf '\n  cri_handler = "cc"')
   privileged_without_host_devices = true
   pod_annotations = ["io.katacontainers.*"]
 EOT
@@ -207,6 +208,16 @@ function configure_containerd() {
 	done
 }
 
+function configure_kata() {
+	if [ "${CONFIGURE_CC:-}" == "yes" ]; then
+		sed -Ei \
+			-e 's#^image = .+#initrd = "/opt/kata/share/kata-containers/kata-containers-initrd.img"#' \
+			-e 's#^(kernel_params = .+)"#\1 agent.config_file=/etc/kata-config.toml"#' \
+			-e 's#.*service_offload = .+#service_offload = true#' \
+			"/opt/kata/share/defaults/kata-containers/configuration-qemu.toml"
+	fi
+}
+
 function remove_artifacts() {
 	echo "deleting kata artifacts"
 	rm -rf /opt/kata/
@@ -282,6 +293,7 @@ function main() {
 
 			install_artifacts
 			configure_cri_runtime "$runtime"
+			configure_kata
 			kubectl label node "$NODE_NAME" --overwrite katacontainers.io/kata-runtime=true
 			;;
 		cleanup)
-- 
2.27.0
```

**Comment by bpradipt** [comment]
> /cc @bpradipt -- I think it would be good to integrate this with https://github.com/confidential-containers/operator/tree/ccv0-demo/demo/payload-image

@Jakob-Naucke for integrating with the operator demo, we'll need changes to kata-deploy.sh as well to include the containerd settings.

**Comment by Jakob-Naucke** [comment]
> @Jakob-Naucke for integrating with the operator demo, we'll need changes to kata-deploy.sh as well to include the containerd settings.

Changes like the one suggested in my patch? :slightly_smiling_face:

**Comment by Jakob-Naucke** [comment]
@bpradipt ^

**Comment by bpradipt** [comment]
> > @Jakob-Naucke for integrating with the operator demo, we'll need changes to kata-deploy.sh as well to include the containerd settings.
> 
> Changes like the one suggested in my patch? 🙂

Yes, to be specific the below one..

`+$([ "$runtime" == "kata" ] || [ "$runtime" == "kata-qemu" ] && [ "${CONFIGURE_CC:-}" == "yes" ] && printf '\n  cri_handler = "cc"')`

Can you please also add a runtimeclass `kata-cc` and ensure the CC related config are added for `kata-cc`. In the operator we explicitly use `kata-cc` as the runtimeclass having the CC specific config (image offload, API restriction etc..)


On a side note, somehow the notifications are going to `/dev/null` :-) .. Need to fix it on my side. .. Sorry for the delay from my side.

**Comment by Jakob-Naucke** [comment]
Okay, here's an updated version that also does the config. I have tested this locally.
The repository in which the action should be run remains an open question.

/test

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @Jakob-Naucke

**Comment by sameo** [comment]
Thanks @Jakob-Naucke !

**Pull request merged**